firmware-management/swupdate-handlers/arm-handler-common.c

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+*/
+
+#include <errno.h>
+#include <limits.h>
+#include <mntent.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include "util.h"
+#include "arm-handler-common.h"
+
+void *malloc_or_abort(const size_t size)
+{
+    char *dst = malloc(sizeof dst * size);
+    if (!dst)
+    {
+        ERROR("%s", "Failed to allocate memory");
+        abort();
+    }
+
+    return dst;
+}
+
+char *str_copy_to_new(const char *const src)
+{
+    const size_t dst_size = strlen(src) + 1;
+    char *dst = malloc_or_abort(dst_size);
+    return strcpy(dst, src);
+}
+
+int str_endswith(const char *const substr, const char *const fullstr)
+{
+    const size_t substr_len = strlen(substr);
+    const size_t fullstr_len = strlen(fullstr);
+    if (substr_len > fullstr_len) return 1;
+    const size_t endlen = fullstr_len - substr_len;
+    return strcmp(&fullstr[endlen], substr);
+}
+
+char *read_file_to_new_str(const char *const filepath)
+{
+    struct stat stat_buf;
+    if (stat(filepath, &stat_buf) == -1)
+    {
+        ERROR("%s %s: %s", "Failed to stat file", filepath, strerror(errno));
+        return NULL;
+    }
+
+    /* st_size is type off_t which is defined as a signed integer type by the POSIX standard,
+       so its safe to cast to size_t here */
+    char *dst = malloc_or_abort((size_t)stat_buf.st_size + 1);
+    char *return_val;
+    FILE *const fp = fopen(filepath, "r");
+    if (fp == NULL)
+    {
+        ERROR("%s %s: %s", "Failed to open file", filepath, strerror(errno));
+        return_val = NULL;
+        goto clean;
+    }
+
+    fread(dst, 1, (size_t)stat_buf.st_size, fp);
+    if (feof(fp) != 0 || ferror(fp) != 0)
+    {
+        ERROR("%s %s", "Failed to read from file", filepath);
+        return_val = NULL;
+        goto clean;
+    }
+
+    dst[stat_buf.st_size] = '\0';
+    return_val = dst;
+
+clean:
+    if (fp != NULL)
+    {
+        if (fclose(fp) == -1)
+        {
+            ERROR("%s %s", "Failed to close file descriptor for file", filepath);
+            return_val = NULL;
+        }
+    }
+
+    if (return_val == NULL)
+        free(dst);
+
+    return return_val;
+}
+
+int get_mounted_device(char *const dst, const char *const mount_point, const size_t dst_size)
+{
+    int return_val = -1;
+
+    FILE *mtab = setmntent("/etc/mtab", "r");
+    if (!mtab)
+    {
+        ERROR("%s", "Failed to open mtab");
+        return -1;
+    }
+
+    struct mntent *mntent_desc;
+
+    while ((mntent_desc = getmntent(mtab)))
+    {
+        if (strcmp(mount_point, mntent_desc->mnt_dir) == 0)
+        {
+            const size_t mnt_fsname_strlen = strlen(mntent_desc->mnt_fsname);
+            strncpy(dst, mntent_desc->mnt_fsname, dst_size);
+
+            if (dst[mnt_fsname_strlen] != '\0')
```

Comment:
If truncation happened, then `mnt_fsname_strlen >= dst_size` and this read is beyond the end of the `dst` buffer.
```suggestion
            if (dst[dst_size] != '\0')
```
Since `strnlen` pads NUL chars to the end of the destination buffer if there's no truncation, `dst[dst_size] != '\0'` should work as a check for truncation.